### PR TITLE
Add eth_call for smart contract methods 

### PR DESCRIPTION
**Pull request by pi0neerpat**
Updates:

- Call methods on a smart contract using ethers or web3.js
- ABI can be entered as a URL or a single ABI function fragment
- Includes contract code sample

image

### CODE REVIEW

File: src/containers/methodCallContainer.js [hunk 1]
```diff
@@ -1,42 +1,101 @@
...
+    const filteredMethods = getFilteredMethods(abi);
+    const formInputsCopy = formInputs;
+    formInputsCopy[2] = {
+      ...formInputs[2],
+      dropdownOptions: filteredMethods,
+      disabled: abi.length === 1,
+    };
+    setFormInputs(formInputsCopy);
+    if (abi.length === 1) updateURL(filteredMethods[0].value, 2);
+  };
```
Comment (abunsen) [inline src/containers/methodCallContainer.js:86]: Same thing here, we can lift this out of here into some other file, pass in args and return values to be updated in the `useEffect` call.

File: src/containers/methodCallContainer.js [hunk 2]
```diff
@@ -50,6 +109,47 @@ const MethodCallContainer = () => {
...
+  }, [formArgs, currentMethod, formInputs]);
+
+  useEffect(() => {
+    if (!abi) return;
+    onUpdateAbi();
+  }, [abi, formInputs]);
+
+  useEffect(() => {
+    if (!argumentList) return;
+    onUpdateContractMethod();
```
Comment (abunsen) [inline src/containers/methodCallContainer.js:145]: associated w/ prev comment about lift.

File: src/containers/methodCallContainer.js [hunk 3]
```diff
@@ -50,6 +109,47 @@ const MethodCallContainer = () => {
...
+  };
+
+  // Load URL arguments
+  useEffect(() => {
+    loadURL();
+  }, [formArgs, currentMethod, formInputs]);
+
+  useEffect(() => {
+    if (!abi) return;
+    onUpdateAbi();
```
Comment (abunsen) [inline src/containers/methodCallContainer.js:140]: associated w/ prev comment about lift.

**Review by abunsen — CHANGES_REQUESTED** [review]
Hey @pi0neerpat - great work dude! 🎉 I am leaving a hand full of comments for you to resolve but this is really close.

File: src/containers/codeSampleContainer.js [hunk 4]
```diff
@@ -21,7 +22,7 @@ const CodeSampleContainer = () => {
     <CodeSample
       url={atob(web3URL)}
       web3Lib={web3Lib}
-      args={argumentList}
+      args={getContractFriendlyArguments(argumentList, abi)}
```
[the remark this answers is not in the document]
Comment (abunsen) [reply src/containers/codeSampleContainer.js:25]: Answer... it does:

Screen Shot 2020-10-01 at 3 06 00 AM

File: src/components/methodCall.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
-import React from 'react';
+import React, { useEffect } from 'react';
```
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/components/methodCall.js]: Removed

File: src/helpers/contracts.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+};
+
+export const getArgumentsFromMethodId = (methodId) => {
+  const [name, rawArgs] = methodId.split('-');
+  if (!rawArgs) return;
+  const args = rawArgs.split(',');
+  return args.map((arg, index) => {
+    return {
+      type: 'textfield',
+      description: `Argument #${index + 1}`,
```
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/helpers/contracts.js:25]: The type is currently in the placeholder position. I think this is enough context, but can move the type to the label if you'd like
image

File: src/containers/codeSampleContainer.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/containers/codeSampleContainer.js]: Fixed

File: src/helpers/libs/ethers.js [hunk 7]
```diff
@@ -19,6 +21,23 @@ const ethersTemplate = (methodCall, varName, url) => {
...
+  const [address, abi, method, ...restArgs] = args;
+  return `const ethers = require("ethers");
+// OR import ethers from 'ethers';
+
+// HTTP version
+(async () => {
+  const abi = '${abi}'
+  const provider = new ethers.providers.JsonRpcProvider('${url}');
+  const contract = new ethers.Contract('${address}', abi, provider);
+  const response = await contract.functions.${method}(${restArgs});
```
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/helpers/libs/ethers.js]: For the sake of "less complicated, less errors",  using strings for everything in the code sample is the best option. I added `getCodeSampleFriendlyArguments()` to handle this and ensure the quotes are included

File: src/helpers/libs/ethers.js [hunk 8]
```diff
@@ -19,6 +21,23 @@ const ethersTemplate = (methodCall, varName, url) => {
...
 
+// TODO: Add Websocket example?
+const contractTemplate = (url, args) => {
+  const [address, abi, method, ...restArgs] = args;
+  return `const ethers = require("ethers");
+// OR import ethers from 'ethers';
+
+// HTTP version
+(async () => {
+  const abi = '${abi}'
```
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/helpers/libs/ethers.js]: Yes, removed the quotes.

File: src/containers/methodCallContainer.js (hunk 3 above)
Comment (pi0neerpat) [reply src/containers/methodCallContainer.js]: Done

File: src/containers/methodCallContainer.js (hunk 2 above)
Comment (pi0neerpat) [reply src/containers/methodCallContainer.js]: Done

File: src/containers/methodCallContainer.js (hunk 1 above)
Comment (pi0neerpat) [reply src/containers/methodCallContainer.js]: Done

File: src/containers/methodCallContainer.js [hunk 9]
```diff
@@ -1,42 +1,101 @@
...
+    if (!methodId) return;
+    const newFormInputs = getArgumentsFromMethodId(methodId);
+    if (newFormInputs)
+      setFormInputs([
+        ...formInputs.slice(0, 3), // Discard existing method-specific inputs
+        ...newFormInputs,
+      ]);
+    else setFormInputs([...formInputs.slice(0, 3)]);
+  };
+
```
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/containers/methodCallContainer.js]: Done

File: src/helpers/web3Config.js [hunk 10]
```diff
@@ -139,8 +139,7 @@ const Web3RpcCalls = {
     ethers: calls.ethers.default.eth_sendRawTransaction,
   },
   eth_call: {
-    disabled: true,
-    description: '🚫 This method is not YET supported in EtherFlow!',
+    description: 'Calls a specific READ function on a contract',
```
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/helpers/web3Config.js]: Done

File: src/containers/methodCallContainer.js [hunk 11]
```diff
@@ -1,55 +1,134 @@
...
     let url = `/${web3URL}/${web3Lib}/`;
     if (currentMethod) url += `${currentMethod}/`;
-    if (args.length > 0) url += `${joinedArgs}`;
+    if (formInputs.length > 0) url += `${joinedArgs}`;
     navigate(url);
   };
-  const runRequest = (args) => {
+
+  const onUpdateArguments = async (val, index) => {
+    if (currentMethod === CONTRACT_FUNCTION_METHOD) {
```
Comment (abunsen) [inline src/containers/methodCallContainer.js:50]: can we combine this line and the one after?

```javascript
if (currentMethod === CONTRACT_FUNCTION_METHOD && index === 1) {
```

File: src/containers/methodCallContainer.js [hunk 12]
```diff
@@ -1,55 +1,134 @@
...
-          method: 'error',
-          data: ['🚨 Error response:', err],
-        });
       });
+    } catch (err) {
+      logItem({
+        method: 'error',
+        data: ['🚨 Error response:', err],
+      });
+    }
```
Comment (abunsen) [inline src/containers/methodCallContainer.js:87]: Why the refactor from `.catch` to `try/catch`? I believe the latter is used for `async/await`.

File: src/helpers/contracts.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (abunsen) [reply src/helpers/contracts.js:27]: Yea, let's put a sample for each type as the placeholder and then `Argument should be an address` as the label under it.

File: src/helpers/contracts.js [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  } catch (e) {
+    return false;
+  }
+  return true;
+};
+
+export const getFilteredMethods = (abi) => {
+  try {
+    return abi
+      .filter((method) => method.stateMutability === 'view')
```
[the remark this answers is not in the document]
Comment (abunsen) [reply src/helpers/contracts.js:45]: Resurfacing this, what is `method.stateMutability`?

File: src/helpers/contracts.js [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+  if (!args || !types) return null;
+  return types.map((type, index) => {
+    if (type === 'address') return `${args[index]}`;
+    return args[index];
+    // TODO: add other types if necessary
+  });
+};
+
+export const getContractFriendlyArguments = (argumentList, abi) => {
+  let [address, , methodId, ...methodSpecificArgs] = argumentList;
```
[the remark this answers is not in the document]
Comment (abunsen) [reply src/helpers/contracts.js:87]: Seems like this was missed?

File: src/helpers/libs/web3js.js [hunk 15]
```diff
@@ -41,6 +41,22 @@ const web3TraceTemplate = (
...
 
+// TODO: Add Websocket example?
+const contractTemplate = (url, args) => {
+  const [address, abi, method, ...restArgs] = args;
+  return `const Web3 = require("web3");
+// OR import Web3 from 'web3';
+
+// HTTP version
+(async () => {
+  const abi = '${abi}'
```
[the remark this answers is not in the document]
Comment (abunsen) [reply src/helpers/libs/web3js.js:52]: Bump.

**Review by abunsen — CHANGES_REQUESTED** [review]
Hopefully last round of changes.

File: src/helpers/contracts.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/helpers/contracts.js:45]: I'm only showing the Read functions https://solidity.readthedocs.io/en/v0.7.2/contracts.html#view-functions

File: src/helpers/contracts.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/helpers/contracts.js:87]: Update here and above  `getArgumentsFromMethodId` with `_` instead of a space.
Also added  `/* eslint-disable-next-line no-unused-vars*/`

File: src/helpers/contracts.js [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  } catch (e) {
+    console.log(e);
+    return { error: ERROR_MESSAGE_PARSE_ABI };
+  }
+};
+
+export const formatContractArgs = (args, types) => {
+  if (!args || !types) return null;
+  return types.map((type, index) => {
+    if (type === 'address') return `${args[index]}`;
```
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/helpers/contracts.js]: N/a. All argument values are strings w/ quotes (See https://github.com/abunsen/etherflow/pull/1#discussion_r498475720)

File: src/helpers/libs/web3js.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/helpers/libs/web3js.js:52]: Oops! Fixed.

File: src/helpers/libs/web3js.js [hunk 17]
```diff
@@ -41,6 +41,22 @@ const web3TraceTemplate = (
...
+  const web3 = new Web3('${url}');
+  const contract = new web3.eth.Contract(abi, '${address}');
+  const response = await contract.methods.${method}(${restArgs});
```
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/helpers/libs/web3js.js]: Fixed with `getCodeSampleFriendlyArguments`. See https://github.com/abunsen/etherflow/pull/1#discussion_r498475720

File: src/containers/methodCallContainer.js (hunk 11 above)
Comment (pi0neerpat) [reply src/containers/methodCallContainer.js:50]: Done.

File: src/containers/methodCallContainer.js (hunk 12 above)
Comment (pi0neerpat) [reply src/containers/methodCallContainer.js:87]: Was debugging and this is just what I'm used to. I changed it back to .catch()

File: src/helpers/contracts.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (pi0neerpat) [reply src/helpers/contracts.js:28]: - I updated the form to now use named input if provided. They are not commonly used in Solidity, but very helpful when they occur.
- I changed the `Argument` to just the type eg. `address`. I think using the "a/an" makes it a bit confusing for things besides "an address", for example "a int","a byes32" etc.
- I added support for array inputs, and added placeholder text as well. This feature is not likely to be used. If you're aware of contracts that use this feature then we can test it and ensure it works for edge cases.

## Form placeholders

Address with named input
image

Address array with named input
image

String with unnamed input
image

String array with unnamed input
image

## Code samples

Address array and unit256
image

**Review by abunsen — APPROVED** [review]
